java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -105,29 +119,29 @@ abstract class AbstractArcHost(
     // TODO: add lifecycle API for ArcHosts shutting down to cancel running coroutines
     private val scope = CoroutineScope(coroutineContext)
 
-    /** Supports asynchronous [ArcHostContext] serializations. */
-    private val channel: Channel<suspend () -> Unit> = Channel(Channel.UNLIMITED)
-    private val updateArcHostContextScope = CoroutineScope(updateArcHostContextCoroutineContext)
+    /**
+     * Supports asynchronous [ArcHostContext] serializations in observed order.
+     *
+     * TODO:
+     * make the channel per-Arc instead of per-Host for better serialization
+     * performance under multiple running and to-be-run Arcs.
+     */
+    private val contextSerializationChannel: Channel<suspend () -> Unit> =
+        Channel(Channel.UNLIMITED)
+    private val contextSerializationTasks =
+        CoroutineScope(updateArcHostContextCoroutineContext).launch {
+            for (task in contextSerializationChannel) task()
+        }
 
     init {
         initialParticles.toList().associateByTo(particleConstructors, { it.first }, { it.second })
-
-        /**
-         * Asynchronous [ArcHostContext] serializations in observed order in a hot [Flow].
-         * The [channel] will be cancelled when the [updateArcHostContextScope] is cancelled.
-         *
-         * TODO:
-         * make the channel per-Arc instead of per-Host for better serialization
-         * performance under multiple running and to-be-run Arcs.
-         */
-        channel.consumeAsFlow().onEach { it() }.launchIn(updateArcHostContextScope)
     }
 
     /** Wait until all observed context serializations got flushed. */
     private suspend fun drainSerializations() {
-        if (!channel.isClosedForSend) {
+        if (!contextSerializationChannel.isClosedForSend) {
```

Comment:
Is this check needed? The channel never gets closed right?

Reply:
If we close in shutdown it could be, but I have a similar question to below:
If we close the channel in ArcHost shutdown, wouldn't it be an error to arrive here with the channel closed? so we should just let it crash?

Reply:
a failed channel: a closed channel due to exceptions.
so used to add a isClosedFor{Send|Receive} check.

Reply:
This looks to me like it masks potential developer errors.

We have full control over when the channel is closed (currently, only due to `cancel` in the `shutdown` method).

We have full control over when we attempt to send to it (occurs due to an outside call to `startArc`, `stopArc`, or `pause`).

My assertion here is that if we call `shutdown`, there should not be subsequent interaction with the ArcHost that results in sending to the channel.